File: cd_week4/multiproductions.py
import re

# Function to validate the Left-Hand Side (LHS) of a production rule
def is_valid_lhs(lhs):
    # Check if the LHS is a single uppercase letter
    return len(lhs) == 1 and lhs.isupper()

# Function to validate the Right-Hand Side (RHS) with multiple alternatives
def is_valid_rhs(rhs):
    # Split RHS by '/' symbol for alternatives
    alternatives = rhs.split('/')
    
    # Validate each alternative
    for alternative in alternatives:
        if not alternative.islower():  # RHS alternatives must be all lowercase
            return False
    return True

# Function to validate a single production rule
def is_valid_production_rule(rule):
    # Check if the rule contains "->"
    if "->" not in rule:
        return False
    
    # Split the rule into LHS and RHS
    lhs, rhs = rule.split("->", 1)  # Split only at the first "->"
    
    # Validate LHS and RHS
    return is_valid_lhs(lhs) and is_valid_rhs(rhs)

# Main function to read multiple production rules and count valid ones
def main():
    num_valid_rules = 0
    
    print("Enter multiple production rules (one per line). Type 'end' to stop input.")
    
    while True:
        # Read each production rule from the user
        rule = input().strip()
        
        if rule.lower() == 'end':  # Stop when user types "end"
            break
        
        # Validate the production rule
        if is_valid_production_rule(rule):
            num_valid_rules += 1
    
    # Output the total number of valid production rules
    print(f"Total number of valid production rules: {num_valid_rules}")

# Run the main function
if __name__ == "__main__":
    main()

File: cd_week4/multiproductions.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <cctype>
#include <vector>

using namespace std;

// Function to validate the Left-Hand Side (LHS) of a production rule
bool isValidLHS(const string& lhs) {
    // Check if the LHS is a single uppercase letter
    return lhs.length() == 1 && isupper(lhs[0]);
}

// Function to validate the Right-Hand Side (RHS) with multiple alternatives
bool isValidRHS(const string& rhs) {
    stringstream ss(rhs);
    string alternative;

    // Split RHS by '/' symbol for alternatives
    while (getline(ss, alternative, '/')) {
        // Check if each alternative consists only of lowercase letters
        for (char ch : alternative) {
            if (!islower(ch)) {
                return false; // Invalid if any character is not lowercase
            }
        }
    }
    return true;
}

// Function to count the number of alternatives in the RHS
int countAlternatives(const string& rhs) {
    stringstream ss(rhs);
    string alternative;
    int count = 0;

    // Split RHS by '/' and count the alternatives
    while (getline(ss, alternative, '/')) {
        count++;
    }

    return count;
}

// Function to validate the entire production rule
bool isValidProductionRule(const string& input) {
    // Find the position of "->"
    size_t arrowPos = input.find("->");
    
    if (arrowPos == string::npos) {
        // If "->" is not found, it's an invalid rule
        return false;
    }
    
    // Extract left-hand side (LHS) and right-hand side (RHS)
    string lhs = input.substr(0, arrowPos);
    string rhs = input.substr(arrowPos + 2); // Skip the "->"

    // Validate LHS and RHS
    return isValidLHS(lhs) && isValidRHS(rhs);
}

int main() {
    string input;
    int numValidRules = 0;

    cout << "Enter multiple production rules (one per line). Type 'end' to stop input.\n";

    // Keep reading production rules until the user enters "end"
    while (true) {
        getline(cin, input);

        // Stop input when user types "end"
        if (input == "end") {
            break;
        }

        // Validate the production rule
        if (isValidProductionRule(input)) {
            numValidRules++;  // Increment the valid rule count
        }
    }

    // Output the total number of valid production rules
    cout << "Total number of valid production rules: " << numValidRules << endl;

    return 0;
}
